File: companies/didi/interview1.cpp
// 求二维矩阵中联通域的个数

#include<iostream>
#include<queue>

using namespace std;

const int N = 10;
const int M = 12;
bool b_isTraver[N][M];

struct axis
{
    int x;
    int y;
};

char g_field[N][M] =
{
    {'w','.','.','.','.','.','.','.','.','w','w','.'},
    {'.','w','w','w','.','.','.','.','.','w','w','w'},
    {'.','.','.','.','w','w','.','.','.','w','w','.'},
    {'.','.','.','.','.','.','.','.','.','w','w','.'},
    {'.','.','.','.','.','.','.','.','.','w','.','.'},
    {'.','.','w','.','.','.','.','.','.','w','.','.'},
    {'.','w','.','w','.','.','.','.','.','w','w','.'},
    {'w','.','w','.','w','.','.','.','.','.','w','.'},
    {'.','w','.','w','.','.','.','.','.','.','w','.'},
    {'.','.','w','.','.','.','.','.','.','.','w','.'},
};

void get_con(queue<axis> &q){
    while (!q.empty())
    {
        for (int i = -1; i <= 1; ++ i)
        {
            for (int j = -1; j <= 1; ++ j)
            {
                if ( (q.front().x+i >= 0) && (q.front().x+i < N) && (q.front().y+j >= 0) && (q.front().y+j < M) && b_isTraver[q.front().x+i][q.front().y+j] == false )
                {
                    b_isTraver[q.front().x+i][q.front().y+j] = true;

                    if (g_field[q.front().x+i][q.front().y+j] == 'w')
                    {
                        axis tmp;
                        tmp.x = q.front().x+i;
                        tmp.y = q.front().y+j;

                        q.push(tmp);
                    }
                }
            }
        }
        q.pop();
    }
}

int main(){

    // 连通组的个数
    int conn = 0;

    // 初始化标志数组
    int i, j;
    for (i = 0; i < N; ++ i){
        for (j = 0; j < M; ++ j)
            b_isTraver[i][j] = false;
    }

    queue<axis> q;

    for (i = 0; i < N; ++ i){
        for (j = 0; j < M; ++ j){
            if (!b_isTraver[i][j] && g_field[i][j] == 'w'){
                ++ conn;

                cout << i << ' ' << j << endl;
                axis tmp;
                tmp.x = i;
                tmp.y = j;
                q.push(tmp);

                b_isTraver[i][j] = true;
                get_con(q);
            }
        }
    }

    cout << conn << endl;

    return 0;
}

File: companies/didi/interview1.py
from collections import namedtuple
from collections import deque

Point = namedtuple('Point', 'x y')

g_field = [['w','.','.','.','.','.','.','.','.','w','w','.'],
           ['.','w','w','w','.','.','.','.','.','w','w','w'],
           ['.','.','.','.','w','w','.','.','.','w','w','.'],
           ['.','.','.','.','.','.','.','.','.','w','w','.'],
           ['.','.','.','.','.','.','.','.','.','w','.','.'],
           ['.','.','w','.','.','.','.','.','.','w','.','.'],
           ['.','w','.','w','.','.','.','.','.','w','w','.'],
           ['w','.','w','.','w','.','.','.','.','.','w','.'],
           ['.','w','.','w','.','.','.','.','.','.','w','.'],
           ['.','.','w','.','.','.','.','.','.','.','w','.']]

ROWS = len(g_field)
COLS = len(g_field[0])

def valid_point(p):
    if p.x < 0 or p.x >= ROWS or p.y < 0 or p.y >= COLS:
        return False
    else:
        return True

def get_conn(de):
    # de not empty
    while de:
        p = de.pop()
        for i in range(-1, 2):
            for j in range(-1, 2):
                neighbor = Point(p.x+i, p.y+j)
                if valid_point(neighbor) and g_field[neighbor.x][neighbor.y] == 'w':
                    de.append(neighbor)
                    g_field[neighbor.x][neighbor.y] = '.'

def main():
    conn = 0
    de = deque()

    for i in range(ROWS):
        for j in range(COLS):
            if g_field[i][j] == 'w':
                conn += 1

                p = Point(i, j)
                de.append(p)
                g_field[i][j] = '.'

                get_conn(de)

    print(conn)

if __name__ == "__main__":
    main()
    print(g_field)